source/shared/core_conn.cpp

```diff
@@ -246,34 +235,36 @@ sqlsrv_conn* core_sqlsrv_connect( _In_ sqlsrv_context& henv_cp, _In_ sqlsrv_cont
         if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
             r = core_odbc_connect( conn, conn_str, is_pooled );
 
-            if( ! SQL_SUCCEEDED( r ) && core_compare_error_state( conn, r, "IM002" ) ) {
+            if( core_compare_error_state( conn, r, "IM002" ) ) {
                 // sql state IM002 means that the specified ODBC driver is not installed
-                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND, CONNECTION_STRING_DRIVER_NAME[ conn->driver_version ].c_str() ) {
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND, CONNECTION_STRING_DRIVER_NAME[conn->driver_version].c_str() ) {
                     throw core::CoreException();
                 }
             }
         }
         else {
-            for( short i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
-                std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION(i) ];
+            bool done = false;
+            for( short i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST && ! done; ++i ) {
+                std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[i];
                 r = core_odbc_connect( conn, conn_str_driver, is_pooled );
 
-                if( SQL_SUCCEEDED( r ) ) {
-                    // successfully connected!
-                    break;
-                }
-                else if(! core_compare_error_state( conn, r, "IM002" ) ) {
-                    // sql state IM002 means that the specified ODBC driver is not installed
-                    // something went wrong other than missing the ODBC driver
-                    break;
+                if( SQL_SUCCEEDED( r ) ) { 
+                    done = true;
```

Comment:
Consider combining the first and last case together since they do the same thing?
`if( SQL_SUCCEEDED( r ) || !core_compare_error_state( conn, r, "IM002" ) )`

Reply:
sorry I'm not sure what you meant here??

Reply:
both `if `and `else `make `done = true`, you can combine them